sway/border.c

```diff
@@ -190,6 +193,39 @@ static void render_title_bar(swayc_t *view, cairo_t *cr, struct wlc_geometry *b,
 		cairo_set_source_u32(cr, colors->text);
 		pango_printf(cr, config->font, 1, false, "%s", view->name);
 	}
+	// Marks
+	if (config->show_marks && view->marks) {
+		int size = 0;
+		char *marks;
+
+		for (int i = 0; i < view->marks->length; ++i) {
+			if (*(char *)view->marks->items[i] != '_') {
+				size += strlen((char *)view->marks->items[i]) + 2;
+			}
+		}
+		marks = (size ? (char *)malloc(size + 1) : strdup("\0"));
+
+		char *cursor = marks;
+		for (int i = 0; i < view->marks->length; ++i) {
+			char *mark = (char *)view->marks->items[i];
+			if (*mark != '_') {
+				int len = strlen(mark);
+				*cursor = '[';
+				strcpy(cursor + 1, mark);
+				*(cursor + len + 1) = ']';
+				cursor += len + 2;
+				*cursor = '\0';
+			}
+		}
+
+		int width, height;
+		get_text_size(cr, config->font, &width, &height, 1, false, "%s", marks);
+		cairo_move_to(cr, MAX((int)tb->size.w + x - (width + 2), x + 2), y + 2);
+		cairo_set_source_u32(cr, colors->text);
+		pango_printf(cr, config->font, 1, false, "%s", marks);
```

Comment:
Better yet, how about just doing this in a loop and avoiding allocation and string slinging alltogether?